File: src/main/nashorn/net/x3djsonld/data/HelloWorldCommented.js
load('X3Dautoclass.js');

// Javadoc annotations follow, see below for source.
/**
 * <p> Numerous comments added to simple X3D scene example for testing JSON encoding alternatives. </p>
 <p> Related links: <a href="../../../HelloWorldScenes/NeedClassName.java">NeedClassName.java</a> source, <a href="../../../HelloWorldScenes/HelloWorldCommentedIndex.html" target="_top">HelloWorldCommented catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../HelloWorldScenes/HelloWorldCommented.x3d">HelloWorldCommented.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> Numerous comments added to simple X3D scene example for testing JSON encoding alternatives. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 19 December 2015 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 20 October 2019 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> info </i> </td>
			<td> HelloWorldCommented.json is most current, HelloWorldCommentedOriginalEncoding.json and HelloWorldCommentedAlternativeEncoding.json were experimental. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="../../../HelloWorldScenes/HelloWorldCommented.json">HelloWorldCommented.json</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="../../../HelloWorldScenes/HelloWorldCommentedOriginalEncoding.json">HelloWorldCommentedOriginalEncoding.json</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="../../../HelloWorldScenes/HelloWorldCommentedAlternativeEncoding.json">HelloWorldCommentedAlternativeEncoding.json</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://x3dgraphics.com/examples/X3dForAdvancedModeling/HelloWorldScenes/HelloWorldCommented.x3d" target="_blank">https://x3dgraphics.com/examples/X3dForAdvancedModeling/HelloWorldScenes/HelloWorldCommented.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/license.html" target="_blank">https://www.web3d.org/x3d/content/examples/license.html</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToES5.xslt" target="_blank">X3dToES5.xslt</a>
		stylesheet to create ES5 source code from an <code>.x3d</code> scene.
	</p>

	* @author Don Brutzman
 */

function NeedClassName
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
NeedClassName.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {
  this.x3dModel = new X3DObject().setProfile("Immersive").setVersion("3.3")
  .setHead(new headObject()
    .addMeta(new metaObject().setName("title").setContent("HelloWorldCommented.x3d"))
    .addMeta(new metaObject().setName("description").setContent("Numerous comments added to simple X3D scene example for testing JSON encoding alternatives."))
    .addMeta(new metaObject().setName("created").setContent("19 December 2015"))
    .addMeta(new metaObject().setName("modified").setContent("20 October 2019"))
    .addMeta(new metaObject().setName("creator").setContent("Don Brutzman"))
    .addMeta(new metaObject().setName("info").setContent("HelloWorldCommented.json is most current, HelloWorldCommentedOriginalEncoding.json and HelloWorldCommentedAlternativeEncoding.json were experimental."))
    .addMeta(new metaObject().setName("reference").setContent("HelloWorldCommented.json"))
    .addMeta(new metaObject().setName("reference").setContent("HelloWorldCommentedOriginalEncoding.json"))
    .addMeta(new metaObject().setName("reference").setContent("HelloWorldCommentedAlternativeEncoding.json"))
    .addMeta(new metaObject().setName("identifier").setContent("https://x3dgraphics.com/examples/X3dForAdvancedModeling/HelloWorldScenes/HelloWorldCommented.x3d"))
    .addMeta(new metaObject().setName("license").setContent("https://www.web3d.org/x3d/content/examples/license.html"))
    .addMeta(new metaObject().setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit")))
  .setScene(new SceneObject()
    .addComments(" Example scene to illustrate comments interspersed among X3D nodes and fields (XML elements and attributes) ")
    .addComments(" WorldInfo begin ")
    .addChild(new WorldInfoObject().setTitle("Hello world!"))
    .addComments(" WorldInfo complete, Group begin ")
    .addChild(new GroupObject()
      .addComments(" Viewpoint begin ")
      .addChild(new ViewpointObject("ViewUpClose").setDescription("Hello world!").setCenterOfRotation(0.0,-1.0,0.0).setPosition(0.0,-1.0,7.0))
      .addComments(" Viewpoint complete, Transform begin ")
      .addChild(new TransformObject().setRotation(0.0,1.0,0.0,3.0)
        .addComments(" Shape begin ")
        .addChild(new ShapeObject()
          .addComments(" Sphere begin ")
          .setGeometry(new SphereObject())
          .addComments(" Sphere complete, Appearance begin ")
          .setAppearance(new AppearanceObject()
            .addComments(" Material begin ")
            .setMaterial(new MaterialObject("MaterialLightBlue").setDiffuseColor(0.1,0.5,1.0))
            .addComments(" Material complete, ImageTexture begin ")
            .setTexture(new ImageTextureObject("ImageCloudlessEarth").setUrl(new MFStringObject("\"earth-topo.png\" \"earth-topo.jpg\" \"earth-topo-small.gif\" \"https://www.web3d.org/x3d/content/examples/Basic/earth-topo.png\" \"https://www.web3d.org/x3d/content/examples/Basic/earth-topo.jpg\" \"https://www.web3d.org/x3d/content/examples/Basic/earth-topo-small.gif\"")))
            .addComments(" ImageTexture complete "))
          .addComments(" Appearance complete "))
        .addComments(" Shape complete "))
      .addComments(" Transform complete, Transform begin ")
      .addChild(new TransformObject().setTranslation(0.0,-2.0,0.0)
        .addComments(" Shape begin ")
        .addChild(new ShapeObject()
          .addComments(" Text begin ")
          .setGeometry(new TextObject("TextMessage").setString(new MFStringObject("\"Hello\" \"world!\""))
            .addComments(" FontStyle begin ")
            .setFontStyle(new FontStyleObject().setJustify(new MFStringObject("\"MIDDLE\" \"MIDDLE\"")))
            .addComments(" FontStyle complete "))
          .addComments(" Text complete, Appearance begin ")
          .setAppearance(new AppearanceObject()
            .addComments(" Material begin ")
            .setMaterial(new MaterialObject().setUSE("MaterialLightBlue"))
            .addComments(" Material complete "))
          .addComments(" Appearance complete "))
        .addComments(" Shape complete "))
      .addComments(" Transform complete "))
    .addComments(" Group complete "));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return NeedClassName model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var       metaResult = "";
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		// first list informational meta elements of interest
		var metaList = this.getX3dModel().getHead().getMetaList();
		for (var m in metaList) {
			meta = metaList[m];
			if (meta.getName().equals(metaObject.NAME_ERROR) ||
				meta.getName().equals(metaObject.NAME_WARNING) ||
				meta.getName().equals(metaObject.NAME_HINT) ||
				meta.getName().equals(metaObject.NAME_INFO) ||
				meta.getName().equals(metaObject.NAME_TODO))
			{
				metaResult += meta.toStringX3D();
			}
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (metaResult === "" && exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = metaResult;
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new NeedClassName();
		print ("NeedClassName execution self-validation test results: " + testObject.validateSelf());
	}
}
new NeedClassName().main();

File: src/main/quarter/net/coderextreme/HelloWorldCommented.cpp

#include <Inventor/nodes/SoCone.h>
#include <Inventor/nodes/SoBaseColor.h>
#include <Inventor/nodes/SoSeparator.h>
#include <Inventor/nodes/SoShape.h>
#include <Inventor/nodes/SoTransform.h>
#include <Inventor/nodes/SoSphere.h>
#include <Inventor/nodes/SoMaterial.h>
#include <Inventor/misc/SoProtoInstance.h>
#include <Inventor/VRMLnodes/SoVRMLAppearance.h>
#include <Quarter/Quarter.h>
#include <Quarter/QuarterWidget.h>
#include <QApplication>
#include <QMainWindow>

using namespace SIM::Coin3D::Quarter;
int main(int argc, char ** argv) 
{

  QApplication app(argc, argv);
  Quarter::init();

  SoSeparator * root = new SoSeparator;
  root->ref();

  SoBaseColor * col = new SoBaseColor;
  col->rgb = SbColor(1, 1, 0);
  root->addChild(col);

  root->addChild(new SoCone);

  QMainWindow * mainwin = new QMainWindow();

  QuarterWidget * viewer = new QuarterWidget(mainwin);
  viewer->setNavigationModeFile();
  mainwin->setCentralWidget(viewer);
  viewer->setSceneGraph(root);

  mainwin->show();
  app.exec();
  root->unref();
  delete viewer;

  Quarter::clean();
SoSceneManager* SoSceneManager0 = new SoSceneManager();
SoSceneManager0->setProfile(QString("Immersive"));
SoSceneManager0->setVersion(QString("4.0"));
Sohead* Sohead1 = new Sohead();
Someta* Someta2 = new Someta();
Someta2->setName(QString("title"));
Someta2->setContent(QString("HelloWorldCommented.x3d"));
Sohead1->addMeta(*Someta2);

Someta* Someta3 = new Someta();
Someta3->setName(QString("description"));
Someta3->setContent(QString("Numerous comments added to simple X3D scene example for testing JSON encoding alternatives."));
Sohead1->addMeta(*Someta3);

Someta* Someta4 = new Someta();
Someta4->setName(QString("created"));
Someta4->setContent(QString("19 December 2015"));
Sohead1->addMeta(*Someta4);

Someta* Someta5 = new Someta();
Someta5->setName(QString("modified"));
Someta5->setContent(QString("20 October 2019"));
Sohead1->addMeta(*Someta5);

Someta* Someta6 = new Someta();
Someta6->setName(QString("creator"));
Someta6->setContent(QString("Don Brutzman"));
Sohead1->addMeta(*Someta6);

Someta* Someta7 = new Someta();
Someta7->setName(QString("info"));
Someta7->setContent(QString("HelloWorldCommented.json is most current, HelloWorldCommentedOriginalEncoding.json and HelloWorldCommentedAlternativeEncoding.json were experimental."));
Sohead1->addMeta(*Someta7);

Someta* Someta8 = new Someta();
Someta8->setName(QString("reference"));
Someta8->setContent(QString("HelloWorldCommented.json"));
Sohead1->addMeta(*Someta8);

Someta* Someta9 = new Someta();
Someta9->setName(QString("reference"));
Someta9->setContent(QString("HelloWorldCommentedOriginalEncoding.json"));
Sohead1->addMeta(*Someta9);

Someta* Someta10 = new Someta();
Someta10->setName(QString("reference"));
Someta10->setContent(QString("HelloWorldCommentedAlternativeEncoding.json"));
Sohead1->addMeta(*Someta10);

Someta* Someta11 = new Someta();
Someta11->setName(QString("identifier"));
Someta11->setContent(QString("https://x3dgraphics.com/examples/X3dForAdvancedModeling/HelloWorldScenes/HelloWorldCommented.x3d"));
Sohead1->addMeta(*Someta11);

Someta* Someta12 = new Someta();
Someta12->setName(QString("license"));
Someta12->setContent(QString("https://www.web3d.org/x3d/content/examples/license.html"));
Sohead1->addMeta(*Someta12);

Someta* Someta13 = new Someta();
Someta13->setName(QString("generator"));
Someta13->setContent(QString("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"));
Sohead1->addMeta(*Someta13);

SoSceneManager0->setHead(*Sohead1);

SoNode* SoNode14 = new SoNode();
//Example scene to illustrate comments interspersed among X3D nodes and fields (XML elements and attributes)
//WorldInfo begin
SoWorldInfo* SoWorldInfo15 = new SoWorldInfo();
SoWorldInfo15->setTitle(QString("Hello world!"));
SoNode14->addChild(*SoWorldInfo15);

//WorldInfo complete, Group begin
SoGroup* SoGroup16 = new SoGroup();
//Viewpoint begin
SoViewpoint* SoViewpoint17 = new SoViewpoint();
SoViewpoint17->setDEF(QString("ViewUpClose"));
SoViewpoint17->setCenterOfRotation(new float[]{0.0,-1.0,0.0});
SoViewpoint17->setDescription(QString("Hello world!"));
SoViewpoint17->setPosition(new float[]{0.0,-1.0,7.0});
SoGroup16->addChild(*SoViewpoint17);

//Viewpoint complete, Transform begin
SoTransform* SoTransform18 = new SoTransform();
SoTransform18->setRotation(new float[]{0.0,1.0,0.0,3.0});
//Shape begin
SoShape* SoShape19 = new SoShape();
//Sphere begin
//Sphere complete, Appearance begin
//Appearance complete
SoSphere* SoSphere20 = new SoSphere();
SoShape19->setGeometry(*SoSphere20);

SoVRMLAppearance* SoVRMLAppearance21 = new SoVRMLAppearance();
//Material begin
//Material complete, ImageTexture begin
//ImageTexture complete
SoMaterial* SoMaterial22 = new SoMaterial();
SoMaterial22->setDEF(QString("MaterialLightBlue"));
SoMaterial22->setDiffuseColor(new float[]{0.1,0.5,1.0});
SoVRMLAppearance21->addChild(*SoMaterial22);

SoImageTexture* SoImageTexture23 = new SoImageTexture();
SoImageTexture23->setDEF(QString("ImageCloudlessEarth"));
SoImageTexture23->setUrl(new QString[]{QString("earth-topo.png"), QString("earth-topo.jpg"), QString("earth-topo-small.gif"), QString("https://www.web3d.org/x3d/content/examples/Basic/earth-topo.png"), QString("https://www.web3d.org/x3d/content/examples/Basic/earth-topo.jpg"), QString("https://www.web3d.org/x3d/content/examples/Basic/earth-topo-small.gif")}, 6);
SoVRMLAppearance21->addChild(*SoImageTexture23);

SoShape19->addChild(*SoVRMLAppearance21);

SoTransform18->addChild(*SoShape19);

//Shape complete
SoGroup16->addChild(*SoTransform18);

//Transform complete, Transform begin
SoTransform* SoTransform24 = new SoTransform();
SoTransform24->setTranslation(new float[]{0.0,-2.0,0.0});
//Shape begin
SoShape* SoShape25 = new SoShape();
//Text begin
//Text complete, Appearance begin
//Appearance complete
SoText* SoText26 = new SoText();
SoText26->setDEF(QString("TextMessage"));
SoText26->setString(new QString[]{QString("Hello"), QString("world!")}, 2);
//FontStyle begin
//FontStyle complete
SoFontStyle* SoFontStyle27 = new SoFontStyle();
SoFontStyle27->setJustify(new QString[]{QString("MIDDLE"), QString("MIDDLE")}, 2);
SoText26->setFontStyle(*SoFontStyle27);

SoShape25->setGeometry(*SoText26);

SoVRMLAppearance* SoVRMLAppearance28 = new SoVRMLAppearance();
//Material begin
//Material complete
SoMaterial* SoMaterial29 = new SoMaterial();
SoMaterial29->setUSE(QString("MaterialLightBlue"));
SoVRMLAppearance28->addChild(*SoMaterial29);

SoShape25->addChild(*SoVRMLAppearance28);

SoTransform24->addChild(*SoShape25);

//Shape complete
SoGroup16->addChild(*SoTransform24);

//Transform complete
SoNode14->addChild(*SoGroup16);

//Group complete
SoSceneManager0->setSceneGraph(*SoNode14);

return 0;
}
